src/app/security/open-id-connect-authorize/open-id-connect-authorize.component.ts

```diff
@@ -87,45 +88,73 @@ export class OpenIdConnectAuthorizeComponent implements OnInit {
     // straight after this
     const sdeProviderName = localStorage.getItem('sdeOpenIdConnectProviderName');
 
-    this.security
-      .authorizeOpenIdConnectSession({
-        providerId,
-        state,
-        sessionState,
-        code,
-      })
-      .pipe(
-        catchError((error: LoginError) => {
-          switch (error.type) {
-            case SignInErrorType.InvalidCredentials:
-              this.errorMessage = 'Invalid username or password!';
-              break;
-            case SignInErrorType.AlreadySignedIn:
-              this.errorMessage = 'A user is already signed in, please sign out first.';
-              break;
-            default:
-              this.errorMessage = 'Unable to sign in. Please try again later.';
-              break;
-          }
-
-          return EMPTY;
-        }),
-        finalize(() => (this.authorizing = false))
-      )
-      .subscribe((user) => {
-        if (!sdeProviderName) {
-          this.broadcast.userSignedIn(user);
-          this.stateRouter.navigateToKnownPath('/dashboard');
-          return;
-        }
-
-        // Auto single sign-on to SDE. There is a big assumption that the OpenID Connect
-        // provider that just succeeded for Mauro is configured exactly the same as for this
-        // SDE provider, so that the sign-in is seamless
-        const redirectUrl =
-          this.authenticationEndpoints.getOauthAuthorizationUrl(sdeProviderName);
-
-        window.open(redirectUrl.toString(), '_self');
+    if (!state || !sessionState || !code) {
+      forkJoin({
+        user: this.catalogueUserService.get('currentUser').pipe(
+          catchError(() => {
+            this.authorizing = false;
+            this.errorMessage = 'OpenID Connect session state has not been provided, and can\'t authenticate current user.';
+            return EMPTY;
+          })
+        ),
+        folder: this.researchPlugin.userFolder()
+      }).subscribe(({user, folder}) => {
+        const userDetails = {
+          id: user.id,
+          email: user.emailAddress,
+          firstName: user.firstName,
+          lastName: user.lastName,
+          role: user.jobTitle
+        } as UserDetails;
+        userDetails.dataSpecificationFolder = folder;
+        this.userDetails.set(userDetails);
+        this.afterUserSignedIn(userDetails, sdeProviderName);
       });
+    } else {
+      this.security
+        .authorizeOpenIdConnectSession({
+          providerId,
+          state,
+          sessionState,
+          code,
+        })
+        .pipe(
+          catchError((error: LoginError) => {
+            switch (error.type) {
+              case SignInErrorType.InvalidCredentials:
+                this.errorMessage = 'Invalid username or password!';
```

Comment:
I'm not sure how this would occur? If you have invalid credentials you cannot log into the OpenID connect provider, so you are not logged in and unable to get here. Unless there is another way to get to this code?